source/common/upstream/load_balancer_impl.cc

```diff
@@ -20,61 +35,77 @@ static const std::string RuntimePanicThreshold = "upstream.healthy_panic_thresho
 LoadBalancerBase::LoadBalancerBase(const PrioritySet& priority_set,
                                    const PrioritySet* local_priority_set, ClusterStats& stats,
                                    Runtime::Loader& runtime, Runtime::RandomGenerator& random)
-    : stats_(stats), runtime_(runtime), random_(random),
-      host_set_(*priority_set.hostSetsPerPriority()[0]),
-      local_host_set_(local_priority_set ? local_priority_set->hostSetsPerPriority()[0].get()
-                                         : nullptr) {
-  if (local_host_set_) {
-    host_set_.addMemberUpdateCb([this](uint32_t, const std::vector<HostSharedPtr>&,
-                                       const std::vector<HostSharedPtr>&) -> void {
-      regenerateLocalityRoutingStructures();
-    });
-    local_host_set_member_update_cb_handle_ =
-        local_host_set_->addMemberUpdateCb([this](uint32_t, const std::vector<HostSharedPtr>&,
-                                                  const std::vector<HostSharedPtr>&) -> void {
-          regenerateLocalityRoutingStructures();
+    : stats_(stats), runtime_(runtime), random_(random), priority_set_(priority_set),
+      best_available_host_set_(bestAvailable(&priority_set)),
+      local_priority_set_(local_priority_set) {
+  resizePerPriorityState(priority_set.hostSetsPerPriority().size());
+  priority_set_.addMemberUpdateCb([this](uint32_t priority, const std::vector<HostSharedPtr>&,
+                                         const std::vector<HostSharedPtr>&) -> void {
+    // Update the host set to use for picking, based on the new state.
+    best_available_host_set_ = bestAvailable(&priority_set_);
+    // If there's a local priority set, regenerate all routing. If not, make
+    // sure per_priority_state_ is still large enough to not cause problems.
```

Comment:
Can you elaborate in this comment on why we need to only regenerate all routing in this situation?